### PR TITLE
Refactor search reduce in QueryNode, remove translateHits and redundant memory copy

### PR DESCRIPTION
**Pull request by bigsheeper**
Signed-off-by: bigsheeper <yihao.dai@zilliz.com>

### CODE REVIEW

File: internal/core/src/segcore/reduce_c.cpp [hunk 1]
```diff
@@ -154,6 +156,208 @@ ReduceResultData(std::vector<SearchResult*>& search_results, int64_t nq, int64_t
...
+        searchResultDataBlob->proto_blob = cSearchResultDataBlobs->blobs[blob_index].proto_blob;
+        searchResultDataBlob->proto_size = cSearchResultDataBlobs->blobs[blob_index].proto_size;
+        return milvus::SuccessCStatus();
+    } catch (std::exception& e) {
+        return milvus::FailureCStatus(UnexpectedError, e.what());
+    }
+}
+
+void
+DeleteSearchResultDataBlobs(CSearchResultDataBlobs* cSearchResultDataBlobs) {
```
Comment (czs007) [inline internal/core/src/segcore/reduce_c.cpp:351]: Write custom destructors for struct cSearchResultDataBlobs!

File: internal/core/src/segcore/reduce_c.cpp [hunk 2]
```diff
@@ -154,6 +156,208 @@ ReduceResultData(std::vector<SearchResult*>& search_results, int64_t nq, int64_t
...
+                                                   slice_offsets[i], slice_offsets[i + 1], output_fields_meta);
+            blobs[i] = cProto;
+        }
+
+        // set final result ptr
+        cSearchResultDataBlobs->num_blobs = num_slices;
+        cSearchResultDataBlobs->blobs = blobs;
+        return milvus::SuccessCStatus();
+    } catch (std::exception& e) {
+        return milvus::FailureCStatus(UnexpectedError, e.what());
```
Comment (czs007) [inline internal/core/src/segcore/reduce_c.cpp:332]: memory leaked when execution occurs.

File: internal/core/src/segcore/reduce_c.cpp [hunk 3]
```diff
@@ -154,6 +156,208 @@ ReduceResultData(std::vector<SearchResult*>& search_results, int64_t nq, int64_t
...
+GetSearchResultDataBlob(CProto* searchResultDataBlob,
+                        CSearchResultDataBlobs* cSearchResultDataBlobs,
+                        int32_t blob_index) {
+    try {
+        AssertInfo(blob_index < cSearchResultDataBlobs->num_blobs, "blob_index out of range");
+        searchResultDataBlob->proto_blob = cSearchResultDataBlobs->blobs[blob_index].proto_blob;
+        searchResultDataBlob->proto_size = cSearchResultDataBlobs->blobs[blob_index].proto_size;
+        return milvus::SuccessCStatus();
+    } catch (std::exception& e) {
+        return milvus::FailureCStatus(UnexpectedError, e.what());
```
Comment (czs007) [inline internal/core/src/segcore/reduce_c.cpp:346]: set  to nullptr and zero when exception occurs

File: internal/core/src/common/SearchResult.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    int64_t topk_;
+    std::vector<float> distances_;
+    std::vector<int64_t> ids_;  // deprecated
+
+ public:
+    // TODO(gexi): utilize these fields
+    void* segment_;
+    std::vector<int64_t> result_offsets_;
+    std::vector<int64_t> primary_keys_;
+    std::vector<std::vector<char>> row_data_;  // deprecated
```
Comment (czs007) [inline internal/core/src/common/SearchResult.h:78]: delete it

File: internal/core/src/segcore/SegmentInterface.cpp [hunk 5]
```diff
@@ -62,36 +62,41 @@ SegmentInternalInterface::FillTargetEntry(const query::Plan* plan, SearchResult&
...
+            results.AddField(field_meta.get_name(), field_meta.get_id(), field_meta.get_data_type(),
+                             field_meta.get_dim(), field_meta.get_metric_type());
+        } else {
+            results.AddField(field_meta.get_name(), field_meta.get_id(), field_meta.get_data_type());
         }
-        assert(element_offset == target_sizeof);
-        results.row_data_.emplace_back(std::move(target));
     }
+
+    // deprecated
```
Comment (czs007) [inline internal/core/src/segcore/SegmentInterface.cpp:83]: delete

File: internal/core/src/segcore/SegmentInterface.cpp [hunk 6]
```diff
@@ -62,36 +62,41 @@ SegmentInternalInterface::FillTargetEntry(const query::Plan* plan, SearchResult&
                        "Primary key field is not INT64 type");
             bulk_subscript(key_offset, results.ids_.data(), size, blob.data());
         }
-        blobs.emplace_back(std::move(blob));
-        element_sizeofs.push_back(sizeof(int64_t));
+        results.ids_data_ = std::move(blob);
```
Comment (czs007) [inline internal/core/src/segcore/SegmentInterface.cpp:65]: unnecessary std::move.
if ids_data_ is not in use, just delete all of this block

File: internal/core/src/segcore/reduce_c.h [hunk 7]
```diff
@@ -22,29 +22,57 @@ extern "C" {
...
 ReduceSearchResultsAndFillData(CSearchPlan c_plan, CSearchResult* search_results, int64_t num_segments);
 
 CStatus
-ReorganizeSearchResults(CMarshaledHits* c_marshaled_hits, CSearchResult* c_search_results, int64_t num_segments);
+ReorganizeSearchResults(CMarshaledHits* c_marshaled_hits,
+                        CSearchResult* c_search_results,
+                        int64_t num_segments);  // deprecated
 
 int64_t
-GetHitsBlobSize(CMarshaledHits c_marshaled_hits);
```
Comment (czs007) [inline internal/core/src/segcore/reduce_c.h:38]: Delete all code about CMarshaledHits

File: internal/core/src/segcore/SegmentInterface.cpp (hunk 6 above)
Comment (bigsheeper) [reply internal/core/src/segcore/SegmentInterface.cpp:65]: what do you mean "delete all of this block"?

File: internal/querynode/reduce_test.go [hunk 8]
```diff
@@ -28,36 +29,122 @@ import (
 	"github.com/milvus-io/milvus/internal/proto/milvuspb"
 )
 
-func TestReduce_AllFunc(t *testing.T) {
-	collectionID := UniqueID(0)
-	segmentID := UniqueID(0)
-	collectionMeta := genTestCollectionMeta(collectionID, false)
+//func TestReduce_AllFunc(t *testing.T) {
```
Comment (czs007) [inline internal/querynode/reduce_test.go:32]: Can this test be rewritten? It is better not to lower the test range.

Comment (bigsheeper) [reply internal/querynode/reduce_test.go:32]: no, `FillTargetEntry` would not assign the row_data member, so old test would crash

File: internal/core/src/common/Types.h [hunk 9]
```diff
@@ -61,37 +61,6 @@ constexpr std::false_type always_false{};
...
-    void* segment_;
-    std::vector<int64_t> result_offsets_;
-    std::vector<int64_t> primary_keys_;
-    std::vector<std::vector<char>> row_data_;
-};
-
-using SearchResultPtr = std::shared_ptr<SearchResult>;
-using SearchResultOpt = std::optional<SearchResult>;
-
 struct RetrieveResult {
```
Comment (xige-16) [inline internal/core/src/common/Types.h:64]: why not move retriveResult class to searchResult.h?

File: internal/core/src/segcore/SegmentInterface.cpp (hunk 6 above)
Comment (bigsheeper) [reply internal/core/src/segcore/SegmentInterface.cpp:65]: old filed named `ids_`, ids_data is the new ids field, so yes, it is in use

Comment (czs007) [reply internal/core/src/segcore/SegmentInterface.cpp:65]: > what do you mean "delete all of this block"?

all related code about ids_data.

Comment (czs007) [reply internal/core/src/segcore/SegmentInterface.cpp:65]: what 's the purpose of ids_data?

File: internal/querynode/reduce_test.go (hunk 8 above)
Comment (czs007) [reply internal/querynode/reduce_test.go:32]: > 

I means keep this test ,but rewrite with your new code .

File: internal/core/src/segcore/reduce_c.cpp (hunk 3 above)
Comment (bigsheeper) [reply internal/core/src/segcore/reduce_c.cpp:346]: fixed

File: internal/core/src/segcore/reduce_c.cpp (hunk 2 above)
Comment (bigsheeper) [reply internal/core/src/segcore/reduce_c.cpp:332]: fixed

File: internal/core/src/segcore/reduce_c.cpp (hunk 1 above)
Comment (bigsheeper) [reply internal/core/src/segcore/reduce_c.cpp:351]: fixed

File: internal/core/src/common/SearchResult.h (hunk 4 above)
Comment (bigsheeper) [reply internal/core/src/common/SearchResult.h:78]: if this delete, reorganizeV1 would build failed, and more cgo build failed too, unless we delete all reorganizeV1 dependancy

File: internal/core/src/segcore/SegmentInterface.cpp (hunk 5 above)
Comment (bigsheeper) [reply internal/core/src/segcore/SegmentInterface.cpp:83]: fixed

File: internal/core/src/segcore/SegmentInterface.cpp (hunk 6 above)
Comment (bigsheeper) [reply internal/core/src/segcore/SegmentInterface.cpp:65]: > what 's the purpose of ids_data?

decouple the ids field from row_data/column_data

File: internal/querynode/reduce_test.go (hunk 8 above)
Comment (bigsheeper) [reply internal/querynode/reduce_test.go:32]: got it

File: internal/core/src/common/Types.h (hunk 9 above)
Comment (bigsheeper) [reply internal/core/src/common/Types.h:64]: agree

File: internal/core/src/segcore/reduce_c.cpp (hunk 1 above)
Comment (bigsheeper) [reply internal/core/src/segcore/reduce_c.cpp:351]: cgo doesn't support destructor, which is a C++ feature

File: internal/core/src/segcore/reduce_c.h (hunk 7 above)
Comment (bigsheeper) [reply internal/core/src/segcore/reduce_c.h:38]: deleted

File: internal/core/src/common/SearchResult.h (hunk 4 above)
Comment (bigsheeper) [reply internal/core/src/common/SearchResult.h:78]: deleted

File: internal/core/src/segcore/reduce_c.cpp (hunk 1 above)
Comment (bigsheeper) [reply internal/core/src/segcore/reduce_c.cpp]: wrap SearchResultDataBlobs，and typedef void* CSearchResultDataBlobs

File: internal/core/src/common/QueryResult.h [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    int64_t topk_;
+    std::vector<float> distances_;
+    std::vector<int64_t> ids_;  // deprecated
+
+ public:
+    // TODO(gexi): utilize these fields
+    void* segment_;
+    std::vector<int64_t> result_offsets_;
+    std::vector<int64_t> primary_keys_;
+    std::vector<std::vector<char>> row_data_;  // deprecated
```
Comment (czs007) [inline internal/core/src/common/QueryResult.h:78]: delete

Comment (bigsheeper) [reply internal/core/src/common/QueryResult.h]: deleted

File: internal/core/unittest/test_c_api.cpp [hunk 11]
```diff
@@ -621,90 +593,84 @@ TEST(CApiTest, ReduceRemoveDuplicates) {
...
-
-    DeleteSearchPlan(plan);
-    DeletePlaceholderGroup(placeholderGroup);
-    DeleteSearchResult(res1);
-    DeleteSearchResult(res2);
-    DeleteMarshaledHits(reorganize_search_result);
-    DeleteCollection(collection);
-    DeleteSegment(segment);
-}
+// TEST(CApiTest, ReduceSearchWithExpr) {
```
Comment (czs007) [inline internal/core/unittest/test_c_api.cpp:596]: Delete This.

File: internal/core/src/segcore/SegmentInterface.cpp [hunk 12]
```diff
@@ -62,35 +61,22 @@ SegmentInternalInterface::FillTargetEntry(const query::Plan* plan, SearchResult&
...
```
Comment (czs007) [inline internal/core/src/segcore/SegmentInterface.cpp:64]: Just modify ids_data_ in previous bulk_subscript?

**Comment by bigsheeper** [comment]
run check

File: internal/core/unittest/test_c_api.cpp (hunk 11 above)
Comment (bigsheeper) [reply internal/core/unittest/test_c_api.cpp]: deleted

File: internal/core/src/segcore/SegmentInterface.cpp (hunk 12 above)
Comment (bigsheeper) [reply internal/core/src/segcore/SegmentInterface.cpp]: fixed

**Pull request merged**